Userland/Games/Chess/ChessWidget.h

```diff
@@ -15,7 +15,7 @@
 #include <LibGfx/Bitmap.h>
 
 class ChessWidget final : public GUI::Frame {
-    C_OBJECT(ChessWidget);
+    C_OBJECT_ABSTRACT(ChessWidget)
```

Comment:
What is the purpose of this change? `ChessWidget` is not an abstract class, it's even marked `final`.

Reply:
The macros are not appropriately named. The `C_OBJECT()` macro implements `class_name()`, `construct()` and `try_create()`. `C_OBJECT_ABSTRACT()` only implements `class_name()`, so the class will only be "abstract" in the sense that some method that someone else wants to use (`window->try_set_main_widget<ChessWidget>()` requres `ChessWidget::try_create()`) will not be automagically implemented. It will *not* mark the class as "abstract" in the C++ sense. We use `C_OBJECT_ABSTRACT()` in `ChessWidget` because we want to implement our own `try_create()` method. The problem is that the default `try_create()` (endowed upon us by `C_OBJECT()`) only propagates OOM errors, but does not allow propagation of other errors during construction, but that is exactly what we want to do here. For this purpose, we write our own `try_create()` method.

If you feel that this is very confusing, I can add notes about this so that future readers won't be confused.

Reply:
The macro is not inappropriately named. `C_OBJECT_ABSTRACT()` is intended for use with abstract classes. The fact that you want to use the macro-inserted `try_create` name for something else is a separate issue.

We're currently in a transition period where new userspace construction patterns are being explored and organically updated. For the time being, I'd recommend simply using a different name for your factory function (instead of `try_create`)

Reply:
Actually, what's even the point of `C_OBJECT_ABSTRACT()` for abstract classes? Since `class_name()` is virtual and all the final children use `C_OBJECT()`, how can the call ever end up in the abstract class?

Reply:
The original point of `C_OBJECT_ABSTRACT()` was to avoid inserting `construct()` into classes that cannot be constructed.

Reply:
Yes, that's clear to me. But I don't see why we needed to annotate abstract classes at all, since, as I said, the only thing they gain is `class_name()`, which I don't think can ever be called.

Anyway, as I said, as long as we keep the `C_OBJECT()` call, `try_set_main_widget()` *will* call `try_create()`, which *will* only propagate errors when allocating the class itself (and not when allocating its subobjects) and then go on to call the infallible constructor. It's simply impossible to improve error propagation without writing a custom `try_create()`.

Reply:
As far as I understand, [#11890](https://github.com/SerenityOS/serenity/pull/11890/commits/99a1ace2c6f8ad7be15cae91dd058ad119f84078#diff-06ed9777bf1acbc6c833035cf1bbde290cb14310da2a374cbbfcfbb699c5321fR15) has shown that it *is* legitimate to use `C_OBJECT_ABSTRACT()` as a name for "`C_OBJECT()` without pre-defined factory functions", at least if the usage of `C_OBJECT_ABSTRACT()` is hidden behind another macro. However, I don't think that there would be much of a point to introduce a new macro such as `C_OBJECT_WITHOUT_FACTORY()` if this was simply a clone of / a wrapper around `C_OBJECT_ABSTRACT()` that wouldn't add anything new. I therefore think that the approach I use here is legitimate in its current form.